### PR TITLE
Do not track coroutines with empty coroutine context in DebugProbes

### PR DESCRIPTION
**Pull request by qwwdfsad**
See #3782 for the rationale

### CODE REVIEW

File: kotlinx-coroutines-debug/src/DebugProbes.kt [hunk 1]
```diff
@@ -59,13 +61,31 @@ public object DebugProbes {
...
+     * `true` by default.
      */
     public var enableCreationStackTraces: Boolean
         get() = DebugProbesImpl.enableCreationStackTraces
         set(value) {
             DebugProbesImpl.enableCreationStackTraces = value
         }
 
+    /**
+     * Whether to ignore coroutines that have their context equal to [EmptyCoroutineContext].
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-debug/src/DebugProbes.kt:74]:
```suggestion
     * Whether to ignore coroutines whose context is [EmptyCoroutineContext].
```
Reflects the code better, if nothing else.

File: kotlinx-coroutines-debug/src/DebugProbes.kt [hunk 2]
```diff
@@ -59,13 +61,31 @@ public object DebugProbes {
...
+     *
+     * Coroutines with empty context are considered to be irrelevant for the concurrent's coroutines observability:
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-debug/src/DebugProbes.kt:76]:
```suggestion
     * Coroutines with empty context are considered to be irrelevant for the concurrent coroutines' observability:
```

File: kotlinx-coroutines-debug/src/DebugProbes.kt [hunk 3]
```diff
@@ -59,13 +61,31 @@ public object DebugProbes {
...
+     * - They do not contribute to any concurrent executions
+     * - They do not contribute to the (concurrent) system's liveness and/or deadlocks as no other coroutines might wait for them
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-debug/src/DebugProbes.kt:78]:
```suggestion
     * - They do not contribute to the (concurrent) system's liveness and/or deadlocks, as no other coroutines might wait for them
```

File: kotlinx-coroutines-debug/src/DebugProbes.kt [hunk 4]
```diff
@@ -59,13 +61,31 @@ public object DebugProbes {
...
+     * - The typical usage of such coroutines is a combinator/builder/lookahead parser that can be debugged using more convenient tools.
+     *
+     * `true` by default.
+     */
+    public var ignoreCoroutinesWithEmptyContext: Boolean
+        get() = DebugProbesImpl.ignoreCoroutinesWithEmptyContext
+        set(value) {
+            DebugProbesImpl.ignoreCoroutinesWithEmptyContext = value
+        }
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-debug/src/DebugProbes.kt:87]: It's peculiar that we can't write it like this:
`public var ignoreCoroutinesWithEmptyContext: Boolean by DebugProbesImpl::ignoreCoroutinesWithEmptyContext`

File: kotlinx-coroutines-debug/test/StandardBuildersAreIgnoredTest.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+                yield(1)
+            }
+        }
+        // Start coroutines
+        fromIterator.hasNext()
+        fromSequence.hasNext()
+
+        val coroutines = DebugProbes.dumpCoroutinesInfo()
+        assertEquals(1, coroutines.size)
+    }
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-debug/test/StandardBuildersAreIgnoredTest.kt:30]: I think the effect of setting `ignoreCoroutinesWithEmptyContext` also needs testing.

File: kotlinx-coroutines-debug/src/DebugProbes.kt (hunk 4 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-debug/src/DebugProbes.kt:87]: That's a bummer indeed.
JFTR, it should be possible to do so with *public* objects (we have internal + suppress)

File: kotlinx-coroutines-debug/test/StandardBuildersAreIgnoredTest.kt (hunk 5 above)
Comment (qwwdfsad) [reply kotlinx-coroutines-debug/test/StandardBuildersAreIgnoredTest.kt:30]: Good point, added

File: kotlinx-coroutines-debug/test/StandardBuildersDebugTest.kt [hunk 6]
```diff
@@ -0,0 +1,50 @@
...
+    }
+
+    @Test
+    fun testBuildersCanBeEnabled() = runTest {
+        try {
+            DebugProbes.ignoreCoroutinesWithEmptyContext = false
+            val (b1, b2) = createBuilders()
+            val coroutines = DebugProbes.dumpCoroutinesInfo()
+            assertEquals(3, coroutines.size)
+            assertTrue { b1.hasNext() && b2.hasNext() } // Don't let GC to collect our coroutines until the test is complete
```
Comment (dkhalanskyjb) [inline kotlinx-coroutines-debug/test/StandardBuildersDebugTest.kt:27]:
```suggestion
        assertTrue { b1.hasNext() && b2.hasNext() } // Don't let GC collect our coroutines until the test is complete
    }

    @Test
    fun testBuildersCanBeEnabled() = runTest {
        try {
            DebugProbes.ignoreCoroutinesWithEmptyContext = false
            val (b1, b2) = createBuilders()
            val coroutines = DebugProbes.dumpCoroutinesInfo()
            assertEquals(3, coroutines.size)
            assertTrue { b1.hasNext() && b2.hasNext() } // Don't let GC collect our coroutines until the test is complete
```

**Review by dkhalanskyjb — APPROVED** [review]

**Pull request merged**